utils/scarb-macro-interface/src/stable_abi.rs

```diff
@@ -0,0 +1,80 @@
+use crate::{ProcMacroResult, TokenStream};
+use std::ffi::CString;
+use std::os::raw::c_char;
+
+/// Token stream.
+///
+/// This struct implements FFI-safe stable ABI.
+#[repr(C)]
+#[derive(Debug)]
+pub struct StableTokenStream(pub *mut c_char);
+
+/// Procedural macro result.
+///
+/// This struct implements FFI-safe stable ABI.
+#[repr(C)]
+#[derive(Debug)]
+#[allow(dead_code)]
+pub enum StableProcMacroResult {
+    /// Plugin has not taken any action.
+    Leave,
+    /// Plugin generated TokenStream replacement.
+    Replace(StableTokenStream),
+    /// Plugin ordered item removal.
+    Remove,
+}
+
+impl StableTokenStream {
+    /// Convert to String.
+    ///
+    /// # Safety
+    pub unsafe fn to_string(&self) -> String {
+        if self.0.is_null() {
+            String::default()
+        } else {
+            let cstr = CString::from_raw(self.0);
+            cstr.to_string_lossy().to_string()
+        }
+    }
+
+    /// Convert to native Rust representation.
+    ///
+    /// # Safety
+    pub unsafe fn to_token_stream(self) -> TokenStream {
+        TokenStream::new(self.to_string())
+    }
+
+    /// Convert to FFI-safe representation.
+    ///
+    /// # Safety
+    pub unsafe fn from_token_stream(token_stream: TokenStream) -> Self {
+        let cstr = CString::new(token_stream.to_string()).unwrap();
```

Comment:
```suggestion
        let cstr = CString::new(token_stream.0).unwrap();
```

this should be possible AFAIK because you're doing this from inner module. You're taking ownership of the TokenStream, so this clone operation is avoidable

Reply:
Yes, I just assumed TokenStream will be changed internally as part of the plugin system implementation.